### PR TITLE
[labs/testing] testing package with SSR testing utilities

### LINKED ISSUE #2867: [labs/ssr] Initial setup of SSR fixture package
Including:
* Initial implementation of `fixture` API
* Initial implementation of WTR plugin/middleware

### PR DESCRIPTION
**Pull request by augustjk**
Resolves #2867 

This PR adds a `@lit-labs/testing` package.

The `litSsrPlugin` plugin for Web Test Runner registers the `lit-ssr-render` server command that takes the template and module paths, and renders them within a worker thread in the same manner as eleventy-plugin-lit.

The server command is utilized by `ssrFixture` to be called in the test file supplying lit-html template to render and component definition module path(s).

The README has example usage.

Still to do are to add tests, and a fixture function with client side rendering with the same signature for ease of writing tests across the different rendering methods.

This is still an early draft and I'm looking for feedback on
- The structure of the package and exports.
- Possibly a better way to handle the relative paths of the modules.
- How to test -- I'm currently thinking doing an integration test with a whole sample component, running WTR with Chrome, FF, Safari, with the plugin registered and running tests utilizing all the fixture function. I'm unsure how to unit test this, if necessary.

### CODE REVIEW

File: packages/labs/testing/README.md
Comment (kevinpschaaf) [inline packages/labs/testing/README.md:58]: I forget what we concluded, but I was also expecting a CSR version with the same signature that used the client-side `render()`?

File: packages/labs/testing/src/web-test-runner/worker/worker.ts
Comment (augustjk) [reply packages/labs/testing/src/web-test-runner/worker/worker.ts:24]: Good point. This structure was necessary in the eleventy plugin because it needed multiple render requests with a single initialize. Here, we start a new worker thread for each command invocation so this could be a lot more simplified.

File: packages/labs/testing/README.md
Comment (augustjk) [reply packages/labs/testing/README.md:58]: Yes, we did want that. It's on the todo list.

**Comment by kevinpschaaf** [comment]
> Possibly a better way to handle the relative paths of the modules.

It seems like we'll need to require the `fixture` caller to pass their `import.meta.url` in order to allow them to use relative paths, something like:
```ts
    const el = await ssrFixture(html`<my-element></my-element>`, {
      base: import.meta.url, // path to resolve relative modules against
      modules: ['./my-element.js'], // path to component definition, relative to base (or cwd if not specified)
      hydrate: false, // whether to hydrate the component after loading to document (default: true)
    });
```

On the client, it would just `import(new URL(module[i], base).href)`, and on the server it should be able to work out where to import it from relative to where the server is serving from?

**Comment by kevinpschaaf** [comment]
> How to test -- I'm currently thinking doing an integration test with a whole sample component, running WTR with Chrome, FF, Safari, with the plugin registered and running tests utilizing all the fixture function. I'm unsure how to unit test this, if necessary

Yeah, doing this as an integration test seems like the most bang-for-buck. Seems like you'll want to test a variety of component scenarios, obv a happy case where it all passes, but also some cases where we expect the tests to fail e.g. to ensure we catch when users do SSR-incompatible things. I'd expect main focus to be on the latter, since the `labs/ssr` integration tests themselves focus on all the happy cases.

^^ points out that we'll want to nicely pipe errors raised from `render-lit-html:render()` on the server back through to the client via a rejection on the async fixture; that should allow writing try/catch blocks in the integration test to ensure they fail when expected.

(hmmm, now that I think about it)
> I'd expect main focus to be on the latter, since the SSR tests themselves focus on all the happy cases.

Seems like we might just be able to switch all the `labs/ssr` integration tests to use this and use a normal WTR setup; it would just require all the element definitions in the `LitElement` tests to be moved out to their own modules. But that can happen separately after we get some basic tests for this landed.

File: packages/labs/testing/src/lib/fixtures/csrFixture.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/fixtures/csrFixture.ts:1]: I think we generally use `kebab-case` instead of `camelCase` for our filenames. (I'm also seeing `snake_case` in some packages, but I think `kebab-case` is more common, e.g. see https://github.com/lit/lit/tree/main/packages/lit/src).

File: packages/labs/testing/src/lib/fixtures/csrFixture.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/fixtures/csrFixture.ts:14]: I don't think we want the `{types}` in the jsdoc, since the types should get correctly extracted by vscode and other tools directly from the typescript syntax, and that's a better source of truth.

File: packages/labs/testing/src/lib/fixtures/csrFixture.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/fixtures/csrFixture.ts:16]: In VSCode this annotation style doesn't quite seem to work -- only the `Path to custom element definition modules` part renders when you are typing the parameter. I think if we annotate the `FixtureOption` properties directly on that type signature, it will work better.

File: packages/labs/testing/src/lib/fixtures/csrFixture.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/fixtures/csrFixture.ts:14]: (Same question as above -- does it really need to be a top-level custom element?)

File: packages/labs/testing/src/lib/fixtures/csrFixture.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/fixtures/csrFixture.ts:18]: Base path for resolving relative paths in `modules`?

File: packages/labs/testing/src/lib/fixtures/csrFixture.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/fixtures/csrFixture.ts:27]: This is quite clever! But maybe kind of surprising? I wouldn't realy expect a function to behave differently depending on where it was called from. Is there some precedent for this, like something `@open-wc/testing` also does?

Another idea could be to just always require the base url be manually specified.

File: packages/labs/testing/src/lib/fixtures/csrFixture.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/fixtures/csrFixture.ts:62]: Isn't this the same on Chrome and Firefox too? Doesn't LitElement always schedule first render on a microtask?

File: packages/labs/testing/src/lib/fixtures/fixtureOption.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/fixtures/fixtureOption.ts:7]:
```suggestion
export interface FixtureOptions {
```

File: packages/labs/testing/src/lib/fixtures/fixtureOption.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/fixtures/fixtureOption.ts:8]: Descriptions

File: packages/labs/testing/src/lib/fixtures/ssrFixture.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/fixtures/ssrFixture.ts:89]: expand on why we must use domparser

File: packages/labs/testing/src/lib/litSsrPlugin.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/litSsrPlugin.ts:23]: Suggest early return here

```suggestion
      if (command !== litSsrPluginCommand) {
        return undefined;
      }
```

**Review by aomarks — CHANGES_REQUESTED** [review]
This is looking really great! Pretty minor comments.

File: packages/labs/testing/src/lib/fixtures/csrFixture.ts
Comment (augustjk) [reply packages/labs/testing/src/lib/fixtures/csrFixture.ts:27]: I had originally made the base required but I added this purely as a convenience for users. This should do the same thing as the user providing `import.meta.url` as the base which I felt like was going to be the vast majority of use cases.

I have not seen any precedent for this and this need specifically arose for this use case because we need be able to resolve filepaths provided as a string to dynamically import from a different module/environment. Our `renderModule` from ssr requires `import.meta.url` to be passed in for the same reason.

File: packages/labs/testing/src/lib/fixtures/csrFixture.ts
Comment (augustjk) [reply packages/labs/testing/src/lib/fixtures/csrFixture.ts:62]: I do not know the answer to this but when I ran the test without the `await` it passed for chromium and firefox, just not webkit. Happy to make the comment just genetically mention awaiting for next microtask to ensure contents are rendered.

File: packages/labs/testing/src/lib/fixtures/ssrFixture.ts
Comment (augustjk) [reply packages/labs/testing/src/lib/fixtures/ssrFixture.ts:102]: We'd need to bring in `hydrate` from `lit/experimental-hydrate.js` and pass in the template and container. I am unsure if that alone would be enough to hydrate any nested lit elements. If not, we'd need to walk down and remove the first `defer-hydration` we see as well.

File: packages/labs/testing/src/lib/worker.ts
Comment (augustjk) [reply packages/labs/testing/src/lib/worker.ts:23]: This won't work because `raw` property is readonly. I did try typing out a `SpoofableTemplateStringsArray` which removes the readonly.

Something like
```ts
type SpoofableTemplateStringsArray = {
  -readonly [P in keyof TemplateStringsArray]: TemplateStringsArray[P];
};

// or

interface SpoofableTemplateStringsArray extends TemplateStringsArray {
  raw: readonly string[];
}
```

Comment (aomarks) [reply packages/labs/testing/src/lib/worker.ts:23]: How about:

```ts
(strings as unknown as {raw: TemplateStringsArray['raw']})
```

Comment (augustjk) [reply packages/labs/testing/src/lib/worker.ts:23]: I like that. Typescript seems to also be okay without the `as unknown`.

File: packages/labs/testing/src/lib/fixtures/ssr-fixture.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/fixtures/ssr-fixture.ts:85]: When I do this:

```ts
new DOMParser().parseFromString("<!--before--><p>hello</p><!--after-->", "text/html")
```

The structure I see is:

```html
<!--before-->
<html>
  <head></head>
  <body>
    <p>hello</p>
    <!--after-->
  </body>
</html>
```

So I think it's not exactly that the comment is ignored, it's that the parser is trying to create a well formed HTML document. It probably keeps the leading comment at the top, because it's allowed to be there, but moves the `<p>` into the `<body>`, because that's the expected well-formed location for it. (The specific element actually matters too -- e.g. a `<p>` will get automatically placed in the `<body>`, while a `<meta>` will get placed in the `<head>`).

So I think your solution is fine, but we could consider doing `<body>${rendered}</body>` and then `fragment.body.childNodes`, to avoid the `querySelector`, and be slightly simpler.

**Review by aomarks — APPROVED** [review]
Looks great! Just some minor comments, LGTM otherwise.

File: packages/labs/testing/src/lib/fixtures/csr-fixture.ts
Comment (kevinpschaaf) [inline packages/labs/testing/src/lib/fixtures/csr-fixture.ts:40]: In what cases do we expect this to fail? It's unclear in the readme if users are expected to pass `import.meta.url` or if they can reasonably expect this to work without having to pass that. Maybe clarify?

File: packages/labs/testing/src/lib/fixtures/csr-fixture.ts
Comment (kevinpschaaf) [inline packages/labs/testing/src/lib/fixtures/csr-fixture.ts:56]: Not sure what the open-wc fixture does, but it might be good to wait a `requestAnimationFrame` instead, to ensure the entire tree of components (if there are multiple levels) are done rendering...?

File: packages/labs/testing/src/lib/worker.ts
Comment (kevinpschaaf) [inline packages/labs/testing/src/lib/worker.ts:28]: Not familiar with how workers work... since this is all executed in module scope, if it throws, then we'll get a worker.on('error') callback?

File: packages/labs/testing/src/lib/lit-ssr-plugin.ts
Comment (kevinpschaaf) [inline packages/labs/testing/src/lib/lit-ssr-plugin.ts:22]: Not blocking for this PR, but would love to think about how this could be factored a little differently so that it could be called from e.g. a koa/express middleware as well, so we're not tied directly to web-test-runner. Should be straightforward, I think?

**Review by kevinpschaaf — APPROVED** [review]
Just a few minor suggestions and clarifications; feel free to consider/resolve, but I'm good merging this based on your judgement on those.

File: packages/labs/testing/src/lib/fixtures/csr-fixture.ts
Comment (augustjk) [reply packages/labs/testing/src/lib/fixtures/csr-fixture.ts:56]: https://github.com/open-wc/open-wc/blob/master/packages/testing-helpers/src/elementUpdated.js

open-wc looks for specific promises to await on and falls back to `requestAnimationFrame` if they don't exist. Perhaps we could do something similar and look for `updateComplete` and await that.

File: packages/labs/testing/src/lib/fixtures/ssr-fixture.ts
Comment (augustjk) [reply packages/labs/testing/src/lib/fixtures/ssr-fixture.ts]: Added a `cleanupFixtures` for now.

I haven't seen cases of users creating multiple fixtures in a single test but don't want to discount that it might happen. I don't feel like trying to automatically clean up is important enough to potentially break that particular use case.

File: packages/labs/testing/src/lib/utils.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/utils.ts:13]: I think we should have some tests for `stripExpressionComments` and `stripExpressionMarkers` probably. Can be an issue/follow up PR though.

File: packages/labs/testing/src/lib/worker.ts
Comment (augustjk) [reply packages/labs/testing/src/lib/worker.ts:28]: Yes that's correct. Currently the wtr server command simply fails for those cases so the fixture function call will then throw. We see that happen in the bad-element test.

File: packages/labs/testing/src/lib/utils.ts
Comment (aomarks) [inline packages/labs/testing/src/lib/utils.ts:7]: Do we want to vend this? It's very basic and not specific to Lit.

File: packages/labs/testing/README.md
Comment (aomarks) [inline packages/labs/testing/README.md:34]: I wonder if this should be more specific a module? Like `@lit-labs/testing/web-test-runner-ssr-plugin.js`? If we ever had another web test runner plugin, it might be nice to keep the module graph distinct.

**Review by aomarks — APPROVED** [review]
LGTM pending 3 comments!

File: packages/labs/testing/src/lib/utils.ts
Comment (augustjk) [reply packages/labs/testing/src/lib/utils.ts:7]: Are you asking if we should or should not vend it? Currently only the strip functions are exported directly from the index. The `nextFrame` would be accessible if the user digs into `lib/utils.js` though.

File: packages/labs/testing/README.md
Comment (augustjk) [reply packages/labs/testing/README.md:34]: Hmm the name is kind of getting long but I see your point about wanted to keep it distinct. `@lit-labs/testing/wtr-ssr-plugin.js`? I don't know if `wtr` is a common enough acronym that's as recognizable.

Comment (aomarks) [reply packages/labs/testing/README.md:34]: Either is OK with me.

File: packages/labs/testing/src/lib/utils.ts
Comment (aomarks) [reply packages/labs/testing/src/lib/utils.ts:7]: Oh I see, that's fine.

File: packages/labs/testing/README.md
Comment (augustjk) [reply packages/labs/testing/README.md]: Went with the longer more descriptive name.

File: packages/labs/testing/src/lib/utils.ts
Comment (augustjk) [reply packages/labs/testing/src/lib/utils.ts:13]: Filed https://github.com/lit/lit/issues/3102

File: packages/labs/testing/src/lib/fixtures/csr-fixture.ts
Comment (augustjk) [reply packages/labs/testing/src/lib/fixtures/csr-fixture.ts:43]: I don't anticipate us to ever reach this unless how the stack is generated gets changed (or the regex somehow fails for some browser besides chrome, firefox, webkit) or if web test runner changes the query to something other than `wtr-session-id`.

I've updated the readme to expand more on the `base` option. Leaving it empty should have the same behavior as passing in `import.meta.url`.

**Pull request merged**
